ENT-1235 Add filtering on audit enrollments for reporting

zubair-arbi opened the pull request:
Enable audit data reporting flag `enable_audit_enrollment` of enterprise customers to also be recognized by the Learner Progress reporting workflow.

- Show/Hide audit enrollment in CSV reports depending on enterprise customers data reporting flag `enable_audit_enrollment`
- Use `edx-django-utils` for caching the enterprise reporting configuration for 6 hours.
```from edx_django_utils.cache import TieredCache```

**Remaining Tasks:**
- Tests coverage

**Story History and Details:**
Currently the "Enable audit data reporting" flag is only recognized by the SAP SuccessFactors reporting workflow. The flag needs to also be recognized by the Learner Progress reporting workflow. See ENT‌-1210 for more information.

1. Should not show audits in CSV reports
2. Admin Dashboard: Filter out audit learner for both overview and UI table (just enrollment data)
    1. Audit Enrollment is filtered out
    2. Users with audit enrollment included in user count but not enrolled count
3. Enterprise Customer model (Flag is in django admin)
4. Data API cache enterprise configuration for 6 hours (for reporting)

**Example:**

- Learners who are only enrolled in audit courses will not be included as “enrolled” learners.
- Count of courses a Learner is enrolled in: If enrolled in 3 audit courses - the count for courses enrolled is 0

zubair-arbi commented:
This PR is ready for initial review.

christopappas commented:
@zubair-arbi looks good from what I can see. I wanted to double check though, there is an `enrollment_count` optionally serialized field in the `EnterpriseUser` serializer (https://github.com/edx/edx-enterprise-data/blob/0136418f4626d83d510d5055e979d086a6ee7512/enterprise_data/api/v0/serializers.py#L38).

Should this also filter auditing users?

zubair-arbi commented:
> I think this is a good start, but my understanding of where we would do the filtering is different than what is here. I believe we wanted to filter at the data api layer so that the admin dashboard would also automatically get the enrollment data filtered appropriately, and then so would the reporting job. Here, it seems to be implemented in the reporting job only. I would move most of what you've implemented for the client/caching to the enterprise_data folder where the data api code lives.

So we need to implement filtering in the pipeline `edx-analytics-pipeline`?

zubair-arbi commented:
@christopappas @brittneyexline I have addressed your feedback. Just need clarification regarding if I implemented the filtering on right place, which according to my understanding we needed to add in views.

brittneyexline commented:
@zubair-arbi - We do need the filtering in the data api, but the fetch to get the data that determines if we should filter should also be done in the data api and not as a flag on the endpoint, so that it happens automatically for any caller of the api. We do not need to make any changes in the pipeline.
I would imagine a new filter class here: https://github.com/edx/edx-enterprise-data/blob/master/enterprise_data/filters.py
This filter would call the enterprise api as you currently have it in the enterprise_reporting/ folder. It would be good to consolidate the caching of the call with the call we make for determining permissions here: https://github.com/edx/edx-enterprise-data/blob/master/enterprise_data/permissions.py#L83
As such, I would expect the caching logic to be applied here: https://github.com/edx/edx-enterprise-data/blob/master/enterprise_data/clients.py
Keep in mind that enterprise_data/ is what gets installed in the analytics data api and serves up the enterprise data api, and enterprise_reporting/ is what gets run by a jenkins job that calls that api. We want the logic for determining if we should filter and performing the filtering to all live in enterprise_data/, so that all callers of the api will get the appropriate filtering without having to know if they are configured to filter out audit enrollments or not.
Let me know if you have any other questions!

zubair-arbi commented:
@brittneyexline I have made the changes now working on tests coverage. Can you please give another review on the changes?

brittneyexline commented on enterprise_data/clients.py:
```diff
@@ -24,6 +29,7 @@ class EnterpriseApiClient(EdxRestApiClient):
 
     API_BASE_URL = settings.LMS_BASE_URL + 'enterprise/api/v1/'
     ENTERPRISE_DATA_API_GROUP = 'enterprise_data_api_access'
+    ENTERPRISE_REPORTING_ENDPOINT = 'enterprise_customer_reporting'
```
technically, we aren't hitting the enterprise reporting endpoint, so i would rename this to indicate that we are hitting with_access_to or the general enterprise-customer endpoint.

brittneyexline commented on enterprise_data/filters.py:
```diff
@@ -25,3 +25,26 @@ def filter_queryset(self, request, queryset, view):
+        Filter out queryset for results where enrollment mode is `audit`.
+        """
+        enterprise_id = view.kwargs['enterprise_id']
+        enable_audit_enrollment = request.session[enterprise_id].get('enable_audit_enrollment', True)
```
i wouldn't default this setting to True - it should default to false

brittneyexline commented on enterprise_data/permissions.py:
```diff
@@ -98,8 +98,12 @@ def has_permission(self, request, view):
             enterprise_data = self.get_enterprise_with_access_to(request.auth, request.user, enterprise_in_url)
             if not enterprise_data:
                 request.session['enterprises_with_access'][enterprise_in_url] = False
+                request.session['enable_audit_enrollment'] = True
```
now that we are adding more than just authorization data in the session, we should probably extract this logic somewhere else to populate the session data with relevant enterprise data.
this assignment also looks to be inconsistent with the one below, which is specific to an enterprise, and that is inconsistent with the structure of the access data. moving this to a separate function gives us the opportunity to make this consistent.

zubair-arbi commented:
Rebased onto master and squashed commits from has 7b09b556bbf04d191e9899ccad64b9beb2e3663e

brittneyexline commented on enterprise_data/permissions.py:
```diff
@@ -94,12 +95,19 @@ def has_permission(self, request, view):
             if not enterprise_data:
-                request.session['enterprises_with_access'][enterprise_in_url] = False
+                update_session_with_enterprise_data(
+                    request, enterprise_in_url, enterprises_with_access=False, enable_audit_enrollment=True
```
enable_audit_enrollment should default to False

zubair-arbi commented:
@saleem-latif This PR is ready for final review.

zubair-arbi commented:
@brittneyexline changes are finalized, please review.

zubair-arbi commented:
Merging the PR now.
